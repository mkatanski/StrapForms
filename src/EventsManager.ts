export interface IEventsManager {
  addListener: (eventName: string, cb: TEventListenerMethod) => void
  removeListener: (eventName: string) => void
  clearListeners: () => void
  dispatchEvent: (methodName: string, eData: IEventData) => void
}

export interface IEventData {
  methodArguments: any[]
}

export type TEventListenerMethod = (eventData: IEventData) => void

export type TListenerObject = {
  [key: string]: Array<TEventListenerMethod>,
}

export class EventData implements IEventData {
  private _args: any[]

  constructor(args: any[]) {
    this._args = args
  }

  public get methodArguments(): any[] {
    return this._args
  }
}

export class EventDataAfter extends EventData implements IEventData {
  private _result: any

  constructor(args: any[], result: any) {
    super(args)
    this._result = result
  }

  public get Result(): any {
    return this._result
  }
}

export class EventsManager implements IEventsManager {
  private _listeners: TListenerObject = {}

  public addListener(eventName: string, cb: TEventListenerMethod): void {
    if (!this._listeners[eventName]) {
      this._listeners[eventName] = []
    }
    this._listeners[eventName].push(cb)
  }

  public removeListener(eventName: string): void {
    delete this._listeners[eventName]
  }

  public clearListeners(): void {
    this._listeners = {}
  }

  public dispatchEvent(methodName: string, eData: IEventData): void {
    let listeners

    if (eData instanceof EventData) {
      listeners = this._listeners[`on_before_${methodName}`]
    }

    if (eData instanceof EventDataAfter) {
      listeners = this._listeners[`on_after_${methodName}`]
    }

    if (!listeners) return

    listeners.forEach((listenerMethod) => {
      listenerMethod(eData)
      console.log(eData)
    })
  }
}

export function EventEmitter(target: any, propertyKey: string, descriptor: PropertyDescriptor) {
  var originalMethod = descriptor.value;

  descriptor.value = function (...args: any[]) {
    this.eventsManager && this.eventsManager.dispatchEvent(propertyKey, new EventData(args))
    const result = originalMethod.apply(this, args)
    this.eventsManager && this.eventsManager.dispatchEvent(propertyKey, new EventDataAfter(args, result))

    return result
  }

  return descriptor;
}

