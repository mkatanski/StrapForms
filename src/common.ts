import { ReactNode } from 'react';

export interface IComponentProps {
  className?: string
  name: string
  key?: any
}

export enum StrapEventType {
  onValidateDone
}

export type TEventData = {
  type: StrapEventType
  data: any
}

export enum ValidationResultType {
  Success,
  Error,
  Warning
}

/**
 * Should contain validation result information
 * such as its type and mesage
 */
export type TValidationResult = {
  /**
   * Describe if the result is success, warning or error
   */
  type: ValidationResultType
  /**
   * Optional messages list that should be displayed
   */
  messages?: Array<string>
}

export type TEventCallback = (event: any) => void
