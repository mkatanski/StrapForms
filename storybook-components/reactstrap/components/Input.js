import React, { Component } from 'react'
import T from 'prop-types'
import { StrapInput, StrapInputPropTypes } from '../../../index'
import { Input, InputGroup, InputGroupAddon, FormText } from 'reactstrap'

class InputPure extends Component {
  static propTypes = {
    ...StrapInputPropTypes,
    helpText: T.string,
  }

  static defaultProps = {
    helpText: '',
  }

  get color() {
    const { hasErrors, hasWarnings, touched } = this.props

    if ((hasErrors || hasWarnings) && touched) {
      return hasErrors ? 'danger' : 'warning'
    }

    return 'default'
  }

  get formText() {
    const { hasErrors, hasWarnings, helpText, touched } = this.props
    if (((hasErrors || hasWarnings) && touched) || !helpText) {
      return
    }

    return <FormText color="muted">{helpText}</FormText>
  }

  render() {
    const { input, isValidating } = this.props

    return (
      <div>
        <InputGroup>
          <Input
            {...input}
            state={this.color}
            type={'text'}
          />
          {isValidating && <InputGroupAddon>validating...</InputGroupAddon>}
        </InputGroup>
        {this.formText}
      </div>
    )
  }
}

export default StrapInput(InputPure)
