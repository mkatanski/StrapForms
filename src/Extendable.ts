import { IEventsManager, EventsManager } from './EventsManager'

export interface IExtendable {
  eventsManager: IEventsManager
}

export type TExtensions = {
  eventsManager?: IEventsManager
}

export class Extendable implements IExtendable {
  private _settings: TExtensions

  constructor(settings?: TExtensions) {
    if (settings) this._settings = settings
  }

  public get eventsManager(): IEventsManager {
    return this._settings ? this._settings.eventsManager : null
  }
}
