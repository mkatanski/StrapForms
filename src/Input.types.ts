import { IComponentProps, StrapEventType, TEventCallback } from './common'
import { ReactNode } from "react";

export interface IInputPureProps extends IComponentProps {
  registerInput: (inputName: string, onEvent: TEventCallback) => void,
  onValidate?: (inputName: string, inputValue: any) => void
  validateUntouched?: boolean
  value?: any
  render?: (props: TInputRendererProps) => ReactNode
  onBlurValidation?: any,
  onChangeValidation?: any,
}

export type TInputState = {
  value: any
  isPristine: boolean
  isTouched: boolean
  isValid: boolean
  validationMessages: Array<string>
}

export type TInputRendererProps = {
  className: string,
  onChange: (e: any) => void,
  onBlur: (e: any) => void,
  value: any,
  name: string,
  isPristine: boolean,
  isTouched: boolean,
  isValid: boolean,
  validationMessages: Array<string>
}
