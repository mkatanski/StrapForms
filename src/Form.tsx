import * as React from "react"
import * as PropTypes from 'prop-types'
import { IValidator, SyncValidator, EResultType, ValidationResult, EValidationType } from './Validators'
import { IValidationManager, ValidationManager } from './ValidationManager'
import { IEventsManager, EventsManager } from './EventsManager'
import { TInputData } from './Input'
import { ReactNode } from "react";
import { IComponentProps } from './common'

export interface IFormProps extends IComponentProps {
  children: any
}

export type TFormRendererProps = {
  children: any
  className: string,
  handleFormSubmit: () => {}
}

export type TFormOptions = {
  isPristine: boolean
  isSubmitted: boolean
  isTouched: boolean
  isValid: boolean
}

export type TFormChildContextTypes = {
  registerInput: (inputName: string, data: TInputData) => void,
  deregisterInput: (inputName: string) => void,
  getValidationManager: () => IValidationManager
  getEventManager: () => IEventsManager
}

export type TInputsData = {
  [key: string]: TInputData
}

export class Form extends React.Component<IFormProps> {
  private _validationManager: ValidationManager
  private eventsManager: EventsManager
  private inputsData: TInputsData = {}
  private _options: TFormOptions = {
    isPristine: true,
    isSubmitted: false,
    isTouched: false,
    isValid: false,
  }

  private _rendererProps: TFormRendererProps = {
    children: this.props.children,
    className: this.props.className,
    handleFormSubmit: this.handleFormSubmit,
  }

  static defaultProps: IFormProps = {
    children: {},
    name: '',
  }

  static childContextTypes = {
    strapActions: PropTypes.object,
  }

  constructor(args: any) {
    super(args)
    this.eventsManager = new EventsManager()
    this._validationManager = new ValidationManager({
      eventsManager: this.eventsManager
    })
  }

  componentDidMount() {
    // console.log(this._validationManager)
  }

  public getChildContext() {
    const context: TFormChildContextTypes = {
      registerInput: this.registerInput,
      deregisterInput: this.deregisterInput,
      getValidationManager: () => {
        return this._validationManager
      },
      getEventManager: () => {
        return this.eventsManager
      }
    }

    return {
      strapActions: context,
    }
  }

  private registerInput(inputName: string, data: TInputData) {
    if (!this.inputsData) this.inputsData = {}
    this.inputsData[inputName] = data
  }

  private deregisterInput(inputName: string) {
    delete this.inputsData[inputName]
    if (this._validationManager) {
      this._validationManager.removeValidator(inputName)
    }
  }

  private async handleFormSubmit(): Promise<any> {

  }

  private defaultRenderer(props: TFormRendererProps): ReactNode {
    return (
      <form className={props.className} onSubmit={props.handleFormSubmit}>
        {props.children}
      </form>
    )
  }

  public render(): ReactNode {
    const { componentRenderer, children } = this.props
    const renderMethod = !!componentRenderer ? componentRenderer : this.defaultRenderer

    return renderMethod(this._rendererProps)
  }
}
