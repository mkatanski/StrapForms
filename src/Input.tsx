import * as React from "react"
import * as PropTypes from 'prop-types'
import { ReactNode } from 'react';
import {
  StrapEventType,
  TEventCallback,
  TEventData,
  TValidationResult,
  ValidationResultType
} from './common'

import {
  IInputPureProps,
  TInputRendererProps,
  TInputState,
} from './Input.types'

import { FormContextConsumer } from './Form/Form'

export class InputPure extends React.Component<IInputPureProps> {
  static defaultProps: IInputPureProps = {
    name: '',
    value: '',
    registerInput: (inputName: string, onEvent: TEventCallback) => { },
  }

  state: TInputState = {
    value: this.props.value,
    isPristine: true,
    isTouched: false,
    isValid: false,
    validationMessages: [],
  }

  public componentDidMount() {
    this.props.registerInput(this.props.name, (event: TEventData) => {
      if (event.type === StrapEventType.onValidateDone) {
        this.onValidationComplete(event.data)
      }
    });
  }

  public componentWillUnmount() {
    // deregister component
  }

  public componentWillReceiveProps(nextProps: IInputPureProps) {
    if (nextProps.value !== this.state.value) {
      this.setState({ value: nextProps.value })
    }
  }

  private onValidationComplete(data: TValidationResult) {

    const messages = data.messages || [];

    if (data.type === ValidationResultType.Success) {
      this.setState({ isValid: true, validationMessages: messages });
    }

    if (data.type === ValidationResultType.Error) {
      this.setState({ isValid: false, validationMessages: messages });
    }

    if (data.type === ValidationResultType.Warning) {
      this.setState({ isValid: false, validationMessages: messages });
    }
  }

  private get rendererProps(): TInputRendererProps {
    return {
      name: this.props.name,
      className: this.props.className,
      onChange: this.handleInputChange,
      onBlur: this.handleInputBlur,
      value: this.state && this.state.value,
      isPristine: this.state && this.state.isPristine,
      isTouched: this.state && this.state.isTouched,
      isValid: this.state && this.state.isValid,
      validationMessages: this.state && this.state.validationMessages,
    }
  }

  private runValidation() {
    const { onValidate, name, validateUntouched } = this.props;
    const { value, isTouched } = this.state;
    onValidate && (isTouched || validateUntouched) && onValidate(name, value);
  }

  private handleInputBlur(e: any): void {
    const { onValidate, name } = this.props;
    const value = e.target ? e.target.value : e

    this.setState({ value, isTouched: true }, this.runValidation)
  }

  private handleInputChange(e: any): void {
    const { onValidate, name, validateUntouched } = this.props;
    const value = e.target ? e.target.value : e

    this.setState({ value, isPristine: false }, this.runValidation)
  }

  private defaultRenderer(componentProps: TInputRendererProps): ReactNode {
    return (
      <input
        {...componentProps}
        onChange={componentProps.onChange.bind(this)}
        onBlur={componentProps.onBlur.bind(this)}
      />
    )
  }

  public render(): ReactNode {
    const { render, children } = this.props
    return !!render ?
      render(this.rendererProps) :
      this.defaultRenderer(this.rendererProps)
  }
}

/**
 * Returns Input component with form's context
 * @param props Input properties
 */
export const Input = (props: any) => (
  <FormContextConsumer>
    {context =>
      <InputPure
        {...props}
        registerInput={context.registerInput}
        onValidate={context.validate}
      />
    }
  </FormContextConsumer>
)
