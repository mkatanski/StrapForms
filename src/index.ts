// import { Form } from './Form/Form'
import { Input } from './Input';
import { Form } from './Form/Form';

import { StrapValidator } from './Validation/StrapValidator'

export {
  Form,
  Input,
  // IValidationClass,
  StrapValidator
}
