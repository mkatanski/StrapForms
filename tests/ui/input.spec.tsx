import { configure } from 'enzyme';
import * as Adapter from 'enzyme-adapter-react-16';
import * as TestRenderer from 'react-test-renderer';
configure({ adapter: new Adapter() });

import * as React from 'react';
import * as ReactDOM from "react-dom";
import * as enzyme from 'enzyme';

import { Input } from '../../src/Input';
import { InputPure } from '../../src/Input';
import { StrapEventType, ValidationResultType, TValidationResult  } from '../../src/common'

jest.mock('react-dom');

enum riArg {
  inputName = 0,
  inputState = 1,
  onEvent = 2
}

describe('Input', () => {
  let registerInputMock: any;

  const getPureInput = (props: any): any => enzyme.shallow(
    (<InputPure
        registerInput={registerInputMock}
        {...props}
    />), {}
  );

  const registerInputCalls = () =>
    registerInputMock.mock.calls

  const registerInputCall = (callNumber: number = 0) =>
    registerInputMock.mock.calls[callNumber]

  beforeEach(() => {
    registerInputMock = jest.fn();
  })

  describe('basic checks', () => {
    let component: any;

    beforeEach(() => {
      component = enzyme.shallow(
          (<InputPure
            name='test input'
            registerInput={jest.fn()}
          /> ), {} );
    })

    it('renders with default component', () => {
      expect(component).toMatchSnapshot();
    });

    it('renders with default component and initial value', () => {
      const input = TestRenderer.create(
        (
          <Input name='defaultInput' value='some value' />
        ),
      );
      expect(input).toMatchSnapshot();
    });

    it('renders with custom class name', () => {
      const input = TestRenderer.create(
        (
          <Input name='customClassName' className='customCssClass' />
        ),
      );
      expect(input).toMatchSnapshot();
    });

    it('renders with custom component', () => {
      const input = TestRenderer.create(
        (
          <Input
            name='customRenderer'
            className='customCssClass'
            render={(props: any) => {

              return (
                <div>
                  name: {props.name}
                  className: {props.className}
                </div>
              )
            }}
          />
        ),
      );
      expect(input).toMatchSnapshot();
    })


    it('updates its value on change', () => {
      expect(component.state('isPristine')).toBe(true)
      expect(component.state('isValid')).toBe(false)
      component.instance().handleInputChange('new value')
      component.update()
      expect(component.state('value')).toBe('new value')
      expect(component.state('isPristine')).toBe(false)
      expect(component).toMatchSnapshot();
    })

    it('updates its value on blur', () => {
      expect(component.state('isTouched')).toBe(false)
      expect(component.state('isValid')).toBe(false)
      component.instance().handleInputBlur('new value on blur')
      component.update()
      expect(component.state('value')).toBe('new value on blur')
      expect(component.state('isTouched')).toBe(true)
      expect(component).toMatchSnapshot();
    })

    it('updates its value when prop is changed', () => {
      component.instance().handleInputChange('new value')
      component.update()

      component.setProps({ value: 'changed value' });
      component.update()
      expect(component.state('value')).toBe('changed value')
      expect(component).toMatchSnapshot();
    })
  })

  describe('registerInput method', () => {
      it('calls registerInput context method', () => {
        const input = getPureInput({
          name: 'MyInput',
        })

        expect(registerInputCalls().length).toBe(1)
        expect(registerInputCall()[riArg.inputName]).toBe('MyInput')
      })
  })

  describe('validation', () => {
    it('calls validation method onInputChange', () => {
      const validateMethod = jest.fn();
      const input = getPureInput({
        name: 'MyPureInput',
        onValidate: validateMethod
      })

      input.instance().handleInputChange('some value to validate');
      expect(input.state('value')).toBe('some value to validate');
      expect(validateMethod.mock.calls.length).toBe(0);

      // set input as touched
      input.setState({
        isTouched: true
      })

      input.instance().handleInputChange('some value to validate');
      expect(validateMethod.mock.calls.length).toBe(1);
      expect(validateMethod.mock.calls[0][0]).toBe('MyPureInput');
      expect(validateMethod.mock.calls[0][1]).toBe('some value to validate');
    })

    it('calls validation method onInputChange untouched', () => {
      const validateMethod = jest.fn();
      const input = getPureInput({
        name: 'MyPureInput',
        validateUntouched: true,
        onValidate: validateMethod
      })

      input.instance().handleInputChange('some value to validate');
      expect(validateMethod.mock.calls.length).toBe(1);
      expect(validateMethod.mock.calls[0][0]).toBe('MyPureInput');
      expect(validateMethod.mock.calls[0][1]).toBe('some value to validate');
    })

    it('calls validation method onInputBlur', () => {
      const validateMethod = jest.fn();
      const input = getPureInput({
        name: 'MyPureInput2',
        onValidate: validateMethod
      })

      input.instance().handleInputBlur('some another value to validate');
      expect(input.state('value')).toBe('some another value to validate');
      expect(validateMethod.mock.calls.length).toBe(1);
      expect(validateMethod.mock.calls[0][0]).toBe('MyPureInput2');
      expect(validateMethod.mock.calls[0][1]).toBe('some another value to validate');
    })
  })

  describe('events', () => {
    let eventCallback;
    let input;

    beforeEach(() => {
      input = getPureInput({
        name: 'MyInput',
        registerInput: (name, callback) => {
          eventCallback = callback;
        }
      });
    })

    it('should assign callback function on input registration', () => {
      const onValidationCompleteMock = jest.fn();
      input.instance().onValidationComplete = onValidationCompleteMock;
      const mockCalls = onValidationCompleteMock.mock.calls;
      expect(eventCallback).toBeDefined();
      expect(mockCalls.length).toBe(0);
    })

    it('should call onValidationComplete', () => {
      const onValidationCompleteMock = jest.fn();
      input.instance().onValidationComplete = onValidationCompleteMock;
      const mockCalls = onValidationCompleteMock.mock.calls;

      const validationResult: TValidationResult = {
        type: ValidationResultType.Success,
        message: 'success message'
      }

      eventCallback({
        type: StrapEventType.onValidateDone,
        data: validationResult
      });

      expect(mockCalls.length).toBe(1);
      expect(mockCalls[0][0]).toEqual({
        type: ValidationResultType.Success,
        message: 'success message'
      });
    })

    it('should has valid state when validation is complete', () => {
      const validationResult: TValidationResult = {
        type: ValidationResultType.Success,
      }

      expect(input.state('isValid')).toBe(false);

      eventCallback({
        type: StrapEventType.onValidateDone,
        data: validationResult
      });

      expect(input.state('isValid')).toBe(true);
    })

    it('should has invalid state when validation is complete and result is error', () => {
      const validationResult: TValidationResult = {
        type: ValidationResultType.Error,
      }

      input.setState({ isValid: true });

      expect(input.state('isValid')).toBe(true);

      eventCallback({
        type: StrapEventType.onValidateDone,
        data: validationResult
      });

      expect(input.state('isValid')).toBe(false);
    })

    it('should has invalid state when validation is complete and result is warning', () => {
      const validationResult: TValidationResult = {
        type: ValidationResultType.Warning,
      }

      input.setState({ isValid: true });

      expect(input.state('isValid')).toBe(true);

      eventCallback({
        type: StrapEventType.onValidateDone,
        data: validationResult
      });

      expect(input.state('isValid')).toBe(false);
    })

  })

});
