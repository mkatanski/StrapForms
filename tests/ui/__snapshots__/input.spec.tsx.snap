// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Input basic checks renders with custom class name 1`] = `
<input
  className="customCssClass"
  isPristine={true}
  isTouched={false}
  isValid={false}
  name="customClassName"
  onBlur={[Function]}
  onChange={[Function]}
  validationMessages={Array []}
  value=""
/>
`;

exports[`Input basic checks renders with custom component 1`] = `
<div>
  name: 
  customRenderer
  className: 
  customCssClass
</div>
`;

exports[`Input basic checks renders with default component 1`] = `
<input
  isPristine={true}
  isTouched={false}
  isValid={false}
  name="test input"
  onBlur={[Function]}
  onChange={[Function]}
  validationMessages={Array []}
  value=""
/>
`;

exports[`Input basic checks renders with default component and initial value 1`] = `
<input
  className={undefined}
  isPristine={true}
  isTouched={false}
  isValid={false}
  name="defaultInput"
  onBlur={[Function]}
  onChange={[Function]}
  validationMessages={Array []}
  value="some value"
/>
`;

exports[`Input basic checks updates its value on blur 1`] = `
<input
  isPristine={true}
  isTouched={true}
  isValid={false}
  name="test input"
  onBlur={[Function]}
  onChange={[Function]}
  validationMessages={Array []}
  value="new value on blur"
/>
`;

exports[`Input basic checks updates its value on change 1`] = `
<input
  isPristine={false}
  isTouched={false}
  isValid={false}
  name="test input"
  onBlur={[Function]}
  onChange={[Function]}
  validationMessages={Array []}
  value="new value"
/>
`;

exports[`Input basic checks updates its value when prop is changed 1`] = `
<input
  isPristine={false}
  isTouched={false}
  isValid={false}
  name="test input"
  onBlur={[Function]}
  onChange={[Function]}
  validationMessages={Array []}
  value="changed value"
/>
`;
